File: 2022/03/solution.py
import pathlib


def priority(l):
    if l.islower():
        return ord(l) - 96
    return ord(l) - 65 + 27


def part1(entries):
    sum = 0
    for l in entries:
        n = len(l)
        compartment1 = set(l[: n // 2])
        compartment2 = set(l[n // 2 :])
        common = compartment1 & compartment2
        assert len(common) == 1
        sum += priority(common.pop())

    return sum


def part2(entries):
    sum = 0
    for idx in range(0, len(entries), 3):
        r1, r2, r3 = entries[idx : idx + 3]
        common = set(r1) & set(r2) & set(r3)
        assert len(common) == 1
        sum += priority(common.pop())

    return sum


def main():
    input = pathlib.Path("input.dat")

    entries = list(l for l in input.read_text().split("\n") if l)
    print(part1(entries))
    print(part2(entries))


if __name__ == "__main__":
    main()

File: 2022/03/solution.hs
import Data.List (intersect)
import Data.List.Split (splitOn)
import Data.Char (ord)

main :: IO()

charPriority :: Char -> Int
charPriority c
  | c >= 'a'  = (ord c) - (ord 'a') + 1 
  | otherwise = (ord c) - (ord 'A') + 27 where

compartmentPriority :: String -> Int
compartmentPriority l = res where
  n = length l
  n2 = n `div` 2
  c1 = take n2 l
  c2 = drop n2 l
  common = intersect c1 c2

  res = case common of
    first:rest -> charPriority first
    _ -> 0
  

part1 :: [String] -> Int
part1 ls = sum where
  priorities = map compartmentPriority ls
  sum = foldl (+) 0 priorities


part2 :: [String] -> Int
part2 ls = sum where
  toPriorities :: String -> String -> String -> Int
  toPriorities x y z = p where
    common = x `intersect` y `intersect` z
    p = charPriority (head common)

  helper (x:y:z:rest) = (toPriorities x y z):(helper rest)
  helper _ = []
  sum = foldl (+) 0 (helper ls)

main = do
  let fileName = "input.dat"
  input <- readFile fileName
  print (part1 (lines input))
  print (part2 (lines input))
